### PR TITLE
rewrite std.path.buildPath in terms of chainPath()

### PR DESCRIPTION
**Pull request by WalterBright**
http://dlang.org/phobos/std_path.html#.buildPath

### CODE REVIEW

**Comment by WalterBright** [comment]
Also: const and inout can't be used with ranges, and fixing this enables the 3+ argument version of chainPath() to work.

File: std/path.d [hunk 1]
```diff
@@ -1230,37 +1230,21 @@ immutable(ElementEncodingType!(ElementType!Range))[]
             if (buf.length < neededLength)
                 buf.length = reserve(buf, neededLength + buf.length/2);
         }
-        if (pos > 0)
+        auto r = chainPath(buf[0 .. pos], segment);
+        size_t i;
+        foreach (c; r)
```
Comment (andralex) [inline std/path.d]: awkward loop, we need an algo here

**Comment by CyberShadow** [comment]
This PR broke building the docs:

```
../phobos/std/path.d(2362): Error: pure function 'std.path.absolutePath' cannot call impure function 'std.path.isAbsolute!string.isAbsolute'
../phobos/std/path.d(2362): Error: safe function 'std.path.absolutePath' cannot call system function 'std.path.isAbsolute!string.isAbsolute'
../phobos/std/path.d(2364): Error: pure function 'std.path.absolutePath' cannot call impure function 'std.path.isAbsolute!string.isAbsolute'
../phobos/std/path.d(2364): Error: safe function 'std.path.absolutePath' cannot call system function 'std.path.isAbsolute!string.isAbsolute'
../phobos/std/path.d(1545): Error: pure function 'std.path.buildNormalizedPath!char.buildNormalizedPath' cannot call impure function 'std.path.isAbsolute!(const(char)[]).isAbsolute'
../phobos/std/path.d(1545): Error: 'std.path.isAbsolute!(const(char)[]).isAbsolute' is not nothrow
../phobos/std/path.d(1512): Error: function 'std.path.buildNormalizedPath!char.buildNormalizedPath' is nothrow yet may throw
../phobos/std/datetime.d(28084): Error: template instance std.path.buildNormalizedPath!char error instantiating
make: *** [docs-prerelease.json] Error 1
```

**Comment by WalterBright** [comment]
Hmm. Why would it have passed the unit tests?

**Comment by CyberShadow** [comment]
The autotester does not attempt to build the documentation. Brad said he's not interested in changing that.

I'm working on a documentation autotester.

Fix: https://github.com/D-Programming-Language/phobos/pull/3390

**Comment by WalterBright** [comment]
Thanks for the quick fix. Ah, I see what the problem is. The StdDoc branch doesn't have an implementation, so no inference is done.

**Pull request merged**